doc/source/tutorials/concurrency-consistency.rst

```diff
@@ -0,0 +1,304 @@
+creation/opening there should be a corresponding array closing.
+
+Concurrent writes are achieved by having each thread or process create
+a **separate fragment** for each write operation. The fragment is
+created upon the first write query submission. No synchronization
+is needed across processes and no internal state is shared across
+threads among the write operations and, thus, *no locking is necessary
+at all*. Regarding the concurrent creation of the fragments, thread- and
+process-safety is achieved because each thread/process creates a fragment
+with a *unique name* (recall, using the process/thread UUID and the current
+timestamp). Therefore, there are no conflicts even at the filesystem level.
+
+A *read opearation* is the process of creating a read
+query object and submitting the query (potentially multiple times in
+the case of incomplete queries) until the query is completed. The code
+snippet shown below describes an **atomic write operation**, i.e.,
+similar to writes, a set of functions that must be invoked by a single
+thread in sequence, but multiple threads can still create their own
+such blocks and invoke them in parallel.
+
+.. content-tabs::
+
+   .. tab-container:: cpp
+      :title: C++
+
+      .. code-block:: c++
+
+         // ### An atomic read operation ###
+         // Optional not recommended: Array array(ctx, array_name, TILEDB_READ);
+         Query query(ctx, array);
+         // ... Prepare query object here
+         query.submit();   // This creates the fragment directory
+         // ... Potentially submit multiple times in global writes
+         query.finalize(); // Important only for global writes
+         // If you opened the array in this block: array.close();
+
+Similar to writes you can optionally open the array in the atomic block.
+This function is both process- and thread-safe. During opening
+the array, TileDB loads the array schema and fragment metadata to main
+memory *once*, and share them across all ``Array`` objects referring to
+the same array. Therefore, for the multi-threading case,
+it is highly recommended that you open the array once *outside* the atomic
+block and have all threads create the query on the same array object.
+This is to prevent the scenario where a thread opens the array, then
+closes it before another thread opens the array again, and so on. TileDB
+internally employs a reference-count system, discarding the array schema
+and fragment metadata each time the array is closed and the reference
+count reaches zero (the schema and metadata are typically cached, but
+they still need to be deserialized in the above scenario). Having
+all concurrent queries use the same ``Array`` object eliminates the
+above problem.
+
+Reads in the multi-processing setting are completely independent
+and no locking is required. In the multi-threading scenario, locking
+is employed (through mutexes) only when the queries access the tile cache,
```

Comment:
In a couple of places we talk about locking, including here, but it may be out of place since the locks you are describing are internal and don't matter to the user. It may just be distracting to talk about internal locking.

Reply:
Locking is an issue in other systems. For instance, SciDB locks the entire array when reading/writing, which caused a serious performance issue with concurrent queries (at least in the past). I just want to emphasize here that TileDB does locking in very few places, and when it does it is for a very short period of time.